addons/account/models/partner.py

```diff
@@ -233,47 +233,52 @@ def _get_fiscal_position(self, partner, delivery=None):
         if not partner:
             return self.env['account.fiscal.position']
 
-        # If no "delivery" partner is specified, we assume it will be the "invoicing" partner.
+        # If no "delivery" partner is specified, we assume it will be the "invoicing" partner
         if not delivery:
             delivery = partner
 
         company = self.env.company
 
+        # Partner manually set fiscal position always wins
+        manual_fiscal_position = (
+            delivery.with_company(company).property_account_position_id
+            or partner.with_company(company).property_account_position_id
+        )
+        if manual_fiscal_position:
+            return manual_fiscal_position
+
+        delivery_country = delivery.country_id
+
         # The purpose of this part is to avoid making (lot of) extra queries by using ref on 'base.europe'
         res_model, res_id = self.env['ir.model.data']._xmlid_to_res_model_res_id('base.europe')
         eu_country_group = self.env[res_model].browse(res_id)
         eu_country_codes = set(eu_country_group.country_ids.mapped('code'))
 
-        delivery_country = delivery.country_id
-
-        eu_vat_partner = partner.vat and partner.vat[:2] in eu_country_codes
-        eu_partner = partner.country_code in eu_country_codes
+        eu_fiscal_partner = partner.vat and partner.vat[:2] in eu_country_codes
         eu_delivery = delivery.country_code in eu_country_codes
+        eu_partner = partner.country_code in eu_country_codes
+        intra_eu = eu_fiscal_partner and eu_partner and eu_delivery
+
+        domestic_fiscal_partner = partner.vat and partner.vat[:2] == company.country_code
         domestic_delivery = delivery_country == company.country_id
 
         vat_required = bool(partner.vat) or domestic_delivery
 
-        # If the delivery is within the EU, the partner does not have a valid EU VAT number and is not from the EU,
-        # then assign the company's country as the delivery country and force vat_required to True
+        # There are 2 "corner" cases here:
+        # - The delivery is within the EU, and both the partner's country and fiscal country are outside the EU.
+        # - An intra community transaction where both the company and partner share the same fiscal country.
+        # In these cases, we assign the company's country as the delivery country and force 'vat_required' to True
         # in order to get the domestic FP
-        if eu_delivery and not eu_vat_partner and not eu_partner:
+        if (eu_delivery and not eu_fiscal_partner and not eu_partner) or (intra_eu and domestic_fiscal_partner):
```

Comment:
I'm not sure we should group the two conditions. It is okay in 16.0 as `vat_required = bool(partner.vat) or domestic_delivery`, but in 17.0, we get it by calling `_get_vat_valid`, which will return `False` if it fails the VIES validation. Do we want to force it anyway?